File: sentimental-credit/credit.py
from cs50 import get_int


def main():
    """ Get the card number from user """
    while True:
        card_number = get_int("Number: ")
        if card_number > 0:
            break
    # Check the card type
    card_type = check_type(card_number)

    # If the card is valid, apply Luhn algorithm
    if card_type != "INVALID":
        if luhn_algorithm(card_number):
            print(card_type)
        else:
            print("INVALID")
    else:
        print("INVALID")


def check_type(card_number):
    """ Check type of card (MasterCard, VISA, AMEX) """
    # Convert to string
    card_number = str(card_number)

    # Check if MC
    if (card_number.startswith(("51", "52", "53", "54", "55")) and len(card_number) == 16):
        return "MASTERCARD"
    # Check if VISA
    elif (card_number.startswith("4") and (len(card_number) == 13 or len(card_number) == 16)):
        return "VISA"
    # Check if AMEX
    elif (card_number.startswith(("34", "37")) and len(card_number) == 15):
        return "AMEX"
    # If nothing from above
    return "INVALID"


def luhn_algorithm(card_number):
    """ Apply Luhn's algorithm """
    # Convert CC number and reverse it
    card_number = str(card_number)[::-1]

    # Initialize sum of digits
    sum = 0

    # Loop through each digit in already reversed card number
    for i, digit in enumerate(card_number):
        digit = int(digit)

        # Double every second digit (from right)
        if i % 2 != 0:
            doubled = digit * 2

            # if result in 2 digits number
            if doubled > 9:
                sum += (doubled % 10) + (doubled // 10)
            else:
                sum += doubled
        else:
            # add digit to the sum
            sum += digit

    # check if total sum can be divided by 10
    return sum % 10 == 0


main()

File: credit/credit.c
// Check if Credit Card provided is valid

#include <cs50.h>
#include <stdbool.h>
#include <stdio.h>

int count_digits(long long card_number); // helper function to count digits
bool luhn_apply(long long card_number);  // function to apply luhn algorithm
string card_type(long long card_number); // determining card type

int main(void)
{
    // Get CC number as input
    long long card_number;
    do
    {
        card_number = get_long("Number: "); // get CC number
    }
    while (card_number < 0);

    // Check if number is valid and apply Luhn's
    if (luhn_apply(card_number))
    {
        printf("%s\n", card_type(card_number));
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

// count digits in a number
int count_digits(long long card_number)
{
    int count = 0;
    while (card_number > 0)
    {
        card_number /= 10;
        count++;
    }
    return count;
}

// apply luhn algorithm
bool luhn_apply(long long card_number)
{
    int sum = 0;
    bool alt = false; // flagging every other digit
    // Multiply every other digit by 2 starting from second-to-last
    while (card_number > 0)
    {
        int digit = card_number % 10;
        card_number /= 10;

        if (alt)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit = (digit / 10) + (digit % 10);
            }
        }
        sum += digit;
        alt = !alt; // flag toggle
    }
    return sum % 10 == 0;
}

// check CC type
string card_type(long long card_number)
{
    long long first_two_digits = card_number;
    while (first_two_digits >= 100)
    {
        first_two_digits /= 10;
    }
    if (count_digits(card_number) == 15 && (first_two_digits == 34 || first_two_digits == 37))
    {
        return "AMEX";
    }
    else if (count_digits(card_number) == 16 && (first_two_digits >= 51 && first_two_digits <= 55))
    {
        return "MASTERCARD";
    }
    else if (first_two_digits / 10 == 4 &&
             (count_digits(card_number) == 13 || count_digits(card_number) == 16))
    {
        return "VISA";
    }
    else
    {
        return "INVALID";
    }
}
